### PR TITLE
Cloudwatch logs formatted similar to native Runtime, v2

### PR DESCRIPTION
**Pull request by mnapoli**
This PR starts from #579 and adds tests + adds a few things.

## The problem

To reuse what @deleugpn described in #579: current errors in PHP lambdas appear like this in logs (CloudWatch):

This is hard to read, **and** this is not consistent with official runtimes like NodeJS. NodeJS logs an exception like this:

## This PR

With these changes, here is how logs will look like:

Capture d’écran 2020-03-14 à 12 21 53

In short, the error is formatted in JSON. Looking good!

Other changes:

- the log message is simpler: it is the error/exception message only
  - it no longer contains `Fatal error:` (the log line contains `Invoke Error`, so that's less noise)
  - it no longer contains `Uncaught exception:`: we don't want users to think "oh I need to catch exceptions" -> exceptions should not be caught, instead Lambda will handle them
  - it no longer contains the file + line number: the first line of the stack trace already contains that
- the log line now contains the "Request ID" (or invocation ID), because:
  - it is closer to NodeJS's output
  - it allows us to easily find the logs of a single invocation
  - all this stems from https://github.com/brefphp/logger/issues/3
- I updated tests and added more of them (as well as made them more readable)

Bonus: how the execution result looks in `serverless invoke`:

Capture d’écran 2020-03-14 à 12 19 08

How the logs look in `serverless logs` (yuck…):

Capture d’écran 2020-03-14 à 12 19 56

### CODE REVIEW

File: src/Runtime/LambdaRuntime.php [hunk 1]
```diff
@@ -78,7 +78,7 @@ private function closeReturnHandler(): void
     /**
      * Process the next event.
      *
-     * @param Handler|callable $handler If it is a callable, it takes two parameters, an $event parameter (mixed) and a $context parameter (Context) and must return anything serializable to JSON.
+     * @param Handler|RequestHandlerInterface|callable $handler If it is a callable, it takes two parameters, an $event parameter (mixed) and a $context parameter (Context) and must return anything serializable to JSON.
```
Comment (mnapoli) [inline src/Runtime/LambdaRuntime.php]: Unrelated fix, sorry about the noise!

File: src/Runtime/LambdaRuntime.php [hunk 2]
```diff
@@ -200,27 +200,23 @@ private function sendResponse(string $invocationId, $responseData): void
...
+        ]) . PHP_EOL;
 
         // Send an "error" Lambda response
         $url = "http://{$this->apiUrl}/2018-06-01/runtime/invocation/$invocationId/error";
         $this->postJson($url, [
-            'errorMessage' => $error->getMessage(),
             'errorType' => get_class($error),
-            'stackTrace' => explode(PHP_EOL, $error->getTraceAsString()),
+            'errorMessage' => $error->getMessage(),
+            'stackTrace' => $stackTraceAsArray,
```
Comment (mnapoli) [inline src/Runtime/LambdaRuntime.php]: Just to be certain to have the same order as Node.

**Comment by deleugpn** [comment]
One last detail of benefit that might be worth mentioning: CloudWatch search makes sense.
Before, I used to struggle (not just with bref but with ECS containers as well) due to PHP stackTrace splitting things into multiple lines.
If you search for an error message, you get a single line in CloudWatch, but dont get the context (previous and next lines), so you lose the stack trace. This is because CloudWatch understands line break as a new entry and php stack is split into lines.
By changing it to an array like this PR does, we can search on CloudWatch for a message and we get a hit that contains the entire json entry, including message + stack.

**Comment by mnapoli** [comment]
Thanks for the feedback! Let's get this going then!
